[PropertyAccess] Major performance improvement

dunglas opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #16179 |
| License | MIT |
| Doc PR | n/a |

This PR improves performance of the PropertyAccess component of ~70%.

The two main changes are:
- caching the `PropertyPath` initialization
- caching the guessed access strategy

This is especially important for the `ObjectNormalizer` (Symfony Serializer) and the JSON-LD normalizer ([API Platform](https://api-platform.com)) because they use the `PropertyAccessor` class in large loops (ex: normalization of a list of entities).

Here is the Blackfire comparison: https://blackfire.io/profiles/compare/c42fd275-2b0c-4ce5-8bf3-84762054d31e/graph

The code of the benchmark I've used (with Symfony 2.3 as dependency):

``` php
<?php

require 'vendor/autoload.php';

class Foo
{
    private $baz;
    public $bar;


    public function getBaz()
    {
        return $this->baz;
    }

    public function setBaz($baz)
    {
        $this->baz = $baz;
    }
}

use Symfony\Component\PropertyAccess\PropertyAccess;

$accessor = PropertyAccess::createPropertyAccessor();

$start = microtime(true);

for ($i = 0; $i < 10000; ++$i) {
    $foo = new Foo();
    $accessor->setValue($foo, 'bar', 'Lorem');
    $accessor->setValue($foo, 'baz', 'Ipsum');
    $accessor->getValue($foo, 'bar');
    $accessor->getValue($foo, 'baz');
}

echo 'Time: '.(microtime(true) - $start).PHP_EOL;
```

This PR also adds an optional support for Doctrine cache to keep access information across requests and improve the overall application performance (even outside of loops).

dunglas committed:
[PropertyAccess] Major performance improvement

dupuchba commented:
:+1:

stof commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -202,48 +211,92 @@ private function &readProperty(&$object, $property)
+
+        if (isset($this->readPropertyCache[$key])) {
+            $access = $this->readPropertyCache[$key];
+        } elseif (!$this->cache || !$access = $this->cache->fetch('read'.$key)) {
```
persisting the cache across requests does not make any sense, as your cache key relies on ``spl_object_hash``, which is not shared across requests

dunglas replied:
Good catch, using the class fully qualified name should should work.

iltar replied:
Is it actually caching values per instance::property or per class::property? Because the first is not really a case you can cover across requests.

dunglas replied:
Actually it's by instance (because of `spl_object_hash`) but I suggest to change it by class (it will also improve performances when different objects of the same type are serialized).

stof commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -329,51 +454,16 @@ private function writeProperty(&$object, $property, $singular, $value)
-                    call_user_func(array($object, $methods[0]), $item);
+                    call_user_func(array($object, $access[self::ACCESS_ADDER]), $item);
                 }
+            break;
```
wrong indentation

dunglas committed:
[PropertyAccess] use get_class instead of spl_object_hash

stof commented:
Using the object hash in the cache key can cause issues if your object is garbage collected and the same hash is reused (which is possible as it is based on the memory location of the object, which is available for reuse).

Btw, your benchmark is not relevant for the API platform use case: you are getting the values thousands of times on the same instance. Your API platform loop would be looping over objects, and so use separate instances. And most of your cache would not be usable in this case, because the object hash would be different and so the cache key would be different. Only the PropertyPath cache would enter into effect. I suggest you to update your benchmark.

What about caching only the logic relying on the class definition instead of caching the logic depending on the object ? This means that you would only need to check dynamic properties each time (and do it before calling the magic methods to respect the existing order of checks). But most cases would still benefit from the cache, and it would be shareable between all objects of the same class, making it much more likely to be reused (and making it shareable across requests too)

dunglas commented:
@stof I just changed strategy (updated the PR) before you post.

stof commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -202,48 +211,92 @@ private function &readProperty(&$object, $property)
+
+                $result[self::VALUE] = &$object->$property;
+                $result[self::IS_REF] = true;
+            } elseif (!$classHasProperty && property_exists($object, $property)) {
```
This cannot be cached, as it depends on the instance, not on the class

stof commented:
@dunglas I suggest you to split this PR in 2 separate PRs:

- one caching the PropertyPath instances, which can be ready quickly
- another one caching the access resolution

You really have 2 independent optimizations here

stloyd commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -445,4 +535,23 @@ private function isAccessible(\ReflectionClass $class, $methodName, $parameters)
+            return $this->propertyPathCache[$propertyPath];
+        }
+
+        $this->propertyPathCache[$propertyPath] = new PropertyPath($propertyPath);
```
```php
return $this->propertyPathCache[$propertyPath] = new PropertyPath($propertyPath);

stof commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -23,16 +24,30 @@ class PropertyAccessor implements PropertyAccessorInterface
      * {@link PropertyAccess::createPropertyAccessor()} instead.
      */
-    public function __construct($magicCall = false)
+    public function __construct($magicCall = false, Cache $cache = null)
```
This is a BC break, because we already have a second argument in newer branches, and it is not a Cache instance

dunglas replied:
I don't like setter injection but it seems to be the way to go to keep BC. What do you think?

aitboudad commented:
IMO this should be considered as new feature ?

dunglas commented:
Benchmark and Blackfire profile updated.

dunglas commented:
@stof I'll move the `PropertyPath` caching in another PR but it's a bit annoying because I plan to use Doctrine cache for `PropertyPath` too.

sescandell commented:
Hi,

IMO, your benchmarks is not exhaustive. What about performance impact on "non-similar" objects?

If I remember well, form component is mainly based on PropertyPath one.
Assuming a very long form (I remember a PR about form performance improvement in that situation) with many fields, what would be the impact of these changes caching objects for "nothing"?

Your benchmark perfectly feet for scenarios you described (Serialization), but PropertyPath might be used somewhere else.

stof commented:
@sescandell the impact on form would be increased memory usage (because of the cache being kept), but the only difference in the code is an array key lookup each time, which is O(1), and much faster than all the reflection-based logic.

And if we store the cache in a persistent cache to reuse it across requests, forms would also benefit from it, as a previous request might have warmed up the cache.

stof commented:
@dunglas your Doctrine cache integration is broken currently anyway: it cannot be merged as is as it would break BC due to the second argument being used already in newer branches.
And the Doctrine cache integration should be considered as a new feature and done in 2.8 IMO. Caching things internally in 2.3 is fine, because it is purely internal, but changing the signature in patch releases looks weird.

stof commented:
Btw, I'm not sure that the PropertyAccessor is the right place to put the PropertyPath caching. We have many places instantiating a PropertyPath outside it (for instance the Form and Validator components always create the instance externally and passes it to the PropertyAccessor).

We may want to introduce a PropertyPathFactory in 2.8, with several implementations:

- a base implementation instantiating the PropertyPath
- a caching implementation wrapping another implementation and caching it in memory
- a doctrine cache implementation (in the bridge) using a persistent cache (this also assumes that we make the PropertyPath class serializable, as we will need to serialize it in the cache)

This is already what we do for the ChoiceListFactory in 2.7+ for instance.

Then, any place needing to create a PropertyPath would rely on the factory (making it an optional argument for BC, and creating a ``new CachingPropertyPathFactory(new PropertyPathFactory())`` by default for instance). This would be BC because ``new PropertyPath()`` still works (but you won't get caching if you use it). This would allow to optimize other places using the PropertyPath too.

stof commented:
Btw, looking at the code of the PropertyPath constructor, I found a simpler optimization: the regex used in it can be optimized simply. I will send a PR for it. Then we will see whether your property path cache is still worth it or no.

sescandell commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -202,48 +211,92 @@ private function &readProperty(&$object, $property)
+                $access[self::ACCESS_TYPE] = self::ACCESS_TYPE_METHOD;
+                $access[self::ACCESS_NAME] = $getter;
+            } else {
+                $methods = array($getter, $isser, $hasser, '__get');
```
Same goes here, it might depend on the instance, and not on the class.

stof replied:
@sescandell there is nothing in this condition depending on ``$object`` (except the fact that the case of dynamic properties must have been checked first)

dunglas committed:
Revert Doctrine Cache and PropertyPath optims

dunglas committed:
Don't cache dynamic properties.

dunglas committed:
Fix CS

stof commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -202,48 +213,93 @@ private function &readProperty(&$object, $property)
+                $result[self::IS_REF] = true;
+            } elseif ($this->magicCall && $reflClass->hasMethod('__call') && $reflClass->getMethod('__call')->isPublic()) {
+                // we call the getter and hope the __call do the job
+                $access[self::ACCESS_TYPE] = self::ACCESS_TYPE_METHOD;
```
this is wrong. You need a separate access type for magic methods, because you must check the dynamic properties before calling the magic method (otherwise it changes the current behavior)

dunglas commented:
@stof OK with your proposal.

* Removed Doctrine cache support from this PR (I'll open a new PR against 2.8 or master if it's too late for 2.8)
* Removed PropertyPatch caching (I'll open another PR if the regex optim isn't enough)

stof commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -202,48 +213,93 @@ private function &readProperty(&$object, $property)
+            default:
+                $reflClass = new \ReflectionClass($object);
+
+                if (!$reflClass->hasProperty($property) && property_exists($object, $property)) {
```
the reflection part could be cached too, to avoid redoing it.

stof commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -291,16 +347,85 @@ private function writeProperty(&$object, $property, $singular, $value)
+                } elseif ($classHasProperty && $reflClass->getProperty($property)->isPublic()) {
+                    $access[self::ACCESS_TYPE] = self::ACCESS_TYPE_PROPERTY;
+                    $access[self::ACCESS_NAME] = $property;
+                } elseif (!$classHasProperty && property_exists($object, $property)) {
```
same issue here for the caching. It depends on the instance

dunglas committed:
Fix guess order

dunglas committed:
Fix CS

dunglas commented:
@stof problems you raised should be fixed now

dunglas commented:
AppVeyor errors not related.

sescandell commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -202,48 +215,86 @@ private function &readProperty(&$object, $property)
+                $access[self::ACCESS_TYPE] = self::ACCESS_TYPE_MAGIC;
+                $access[self::ACCESS_NAME] = $getter;
+            } else {
+                $methods = array($getter, $isser, $hasser, '__get');
```
You should not cache the not found case.
Imagine the following situation:
```php
$a = new stdClass();
$a->a = 'a';

$b = new stdClass();
$b->b = 'b';

$accessor->getValue('b', $a);
$accessor->getValue('b', $b);
```

In that situation it will just throw the exception (except that for $b it should work) and if calls were reversed, it will simply crash when trying to access  `b` property from `$a` because of cached path.

dunglas replied:
It will work with your example: https://github.com/dunglas/symfony/blob/propertyaces-perf/src/Symfony/Component/PropertyAccess/PropertyAccessor.php#L284-L293

Only method names are cached but dynamic properties are always tried.

sescandell replied:
:+1:

stof commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -202,48 +215,86 @@ private function &readProperty(&$object, $property)
+        } else {
+            $access = array();
+
+            $access[self::ACCESS_REFLECTION_CLASS] = $reflClass = new \ReflectionClass($object);
```
putting the ReflectionClass in the cached data would make it much harder to persist the cache across requests. I suggest caching ``$classHasProperty`` instead (as this is the only thing you need actually)

stof commented:
@dunglas to make the code more readable, the cache building may be extracted to separate private methods

dunglas committed:
Fix @stof comments

dunglas commented:
@stof done. Did you have the time to take a look at the PropertyPath regex?

dunglas commented:
ping @symfony/deciders

fabpot commented:
Thank you @dunglas.

sstok commented:
I really like Stofs idea for caching the determined strategy, and caching the property-path to object :+1:

dunglas commented:
I removed these features from this PR because they cannot be merged in 2.3 but I'll open PR new PRa against 2.8 yo add them back.

fabpot commented:
@dunglas I've just merged 2.3 into 2.7 without including your modifications as it conflicts badly with all the changes made since 2.3. Instead of trying to merge everything, I think it would be better to just create a new PR on 2.7. Can you take care of that? Thanks.

dunglas commented:
Ok I'll do it this week.

webmozart commented:
This is awesome, thanks @dunglas!!
